File: Modularity-State-Management-Architecture/Library.Infrastructure/Library/StateBuilders/StateBuilderBase.cs
using Library.Infrastructure.Library.StateBuilders;
using StateBuilder.Library.Infrastructure;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Unity;
namespace StateBuilder.Library.StateBuilders
{
    public abstract class StateBuilderBase<TStateObject>
            where TStateObject : class
    {
        private readonly IWorkItem _workItem;
        private readonly string _stateName;

        private StateBuilderContext<TStateObject> _stateBuilderContext;

        protected StateBuilderBase(IWorkItem workItem, string stateName)
        {
            _workItem = workItem;
            _stateName = stateName;
            
            var stateRegistry = workItem.RootWorkItem.Container.Resolve<IStateRegistry>();
            stateRegistry.RegisterStateType(workItem.GetType(), _stateName, typeof(TStateObject));
        }

        public void Run()
        {
            if (GetDependencies().Contains(_stateName))
            {
                throw new InvalidOperationException("StateBuilder cannot have itself as a dependency.");
            }
            if (GetOptionalDependencies().Contains(_stateName))
            {
                throw new InvalidOperationException("StateBuilder cannot have itself as a dependency.");
            }

            _workItem.SubscribeStateChangedEvent(_stateName, OnStateChanged);

            var dependencies = GetDependencies().ToArray();
            foreach (var dependencyName in dependencies)
            {
                _workItem.SubscribeStateChangedEvent(dependencyName, OnDependencyChanged);
            }
            if (!dependencies.Any())
            {
                OnDependencyChanged();
            }

            var optional = GetOptionalDependencies();
            foreach (var dependencyName in optional)
            {
                _workItem.SubscribeStateChangedEvent(dependencyName, OnDependencyChanged);
            }
        }

        protected void OnDependencyChanged()
        {
            var dependencies = GetDependencies().ToDictionary(x => x, x => _workItem.State[x]);
            var optionalDependencies = GetOptionalDependencies().ToDictionary(x => x, x => _workItem.State[x]);

            var stateBuilderContext = CreateStateBuilderContext(dependencies, optionalDependencies);
            if (Interlocked.CompareExchange(ref _stateBuilderContext, stateBuilderContext, null) != null)  // try to assign new state handler
            {
                var e = _stateBuilderContext;  // an existing state handler is running, try to cancel and overwrite with new state handler
                while (Interlocked.CompareExchange(ref _stateBuilderContext, stateBuilderContext, e) != e)
                {
                    e?.Cancel();
                    e = _stateBuilderContext;
                }
                e?.Cancel();
            }

            if (stateBuilderContext.Dependencies.Values.All(x => x != null))
            {
                OnDependenciesMet(stateBuilderContext);
            }
            else
            {
                OnDependenciesNotMet(stateBuilderContext);
            }
        }

        protected void OnDependenciesMet(StateBuilderContext stateBuilderContext)
        {
            Task task = Task.Run(async delegate
            {
                try
                {
                    await Task.Delay(TimeSpan.FromMilliseconds(50), stateBuilderContext.CancellationToken);
                    if (!stateBuilderContext.CancellationToken.IsCancellationRequested)
                    {
                        DoWork(stateBuilderContext);
                    }
                }
                catch (TaskCanceledException)
                {
                    stateBuilderContext.Cancel();
                }
                catch (Exception ex)
                {
                    OnException(ex);
                    stateBuilderContext.Cancel();
                }
                finally
                {
                    stateBuilderContext = null;
                }
            }, stateBuilderContext.CancellationToken);
            var taskList = _workItem.Container.Resolve<IList<Task>>();
            taskList.Add(task);
            /*Task.Run(async delegate
            {
                try
                {
                    await Task.Delay(TimeSpan.FromMilliseconds(50), stateBuilderContext.CancellationToken);
                    if (!stateBuilderContext.CancellationToken.IsCancellationRequested)
                    {
                        DoWork(stateBuilderContext);
                    }
                }
                catch (TaskCanceledException)
                {
                    stateBuilderContext.Cancel();
                }
                catch (Exception ex)
                {
                    OnException(ex);
                    stateBuilderContext.Cancel();
                }
                finally
                {
                    stateBuilderContext = null;
                }
            }, stateBuilderContext.CancellationToken);*/
        }

        protected void OnDependenciesNotMet(StateBuilderContext stateBuilderContext)
        {
            stateBuilderContext.SetResult(null);
        }

        protected virtual IEnumerable<string> GetDependencies()
        {
            return Enumerable.Empty<string>();
        }

        protected virtual IEnumerable<string> GetOptionalDependencies()
        {
            return Enumerable.Empty<string>();
        }

        protected virtual IEnumerable<string> GetAdditionalData()
        {
            return Enumerable.Empty<string>();
        }

        protected void DoWork(StateBuilderContext stateBuilderContext)
        {
            GetStateObject(stateBuilderContext);
        }

        protected abstract void GetStateObject(StateBuilderContext context);

        protected StateBuilderContext CreateStateBuilderContext(IDictionary<string, object> dependencies, IDictionary<string, object> optionalDependencies)
        {
            var additionalData = GetAdditionalData().ToDictionary(x => x, x => _workItem.State[x]);
            return new StateBuilderContext(_workItem, _stateName, dependencies, optionalDependencies, additionalData);
        }


        protected class StateBuilderContext : StateBuilderContext<TStateObject>
        {
            public StateBuilderContext(IWorkItem workItem, string stateName, IDictionary<string, object> dependencies, IDictionary<string, object> optionalDependencies, IDictionary<string, object> additionalData)
                : base(workItem, stateName, dependencies, optionalDependencies, additionalData)
            {
            }
        }

        private void OnStateChanged(StateChangedEventArgs eventArgs)
        {
            if (eventArgs.OldValue != null && eventArgs.NewValue == null)
            {
                OnDependencyChanged();
            }
        }

        protected void OnException(Exception ex)
        {
        }
    }
}

File: Modularity-State-Management-Architecture/WebApplicationNG/ClientApp/src/StateBuilders/Library/StateBuilders/StateBuilderBase.ts
import { IWorkItem } from '../Infrastructure/IWorkItem';
import { Dictionary } from '../../../Abstractions/ViewControllerBase';
import { StateBuilderContext } from './StateBuilderContext';
import { StateChangedEventArgs } from '../Infrastructure/StateChangedEventArgs';

export abstract class StateBuilderBase<TStateObject extends any>{
  public _workItem:IWorkItem;
  public _stateName:string;
  public _stateBuilderContext:StateBuilderContext<any>;

  protected constructor(workItem:IWorkItem,stateName:string){
    this._workItem=workItem;
    this._stateName=stateName;
  }

  public Run():void{
    this.OnRunStarted();
  }

  public OnStateChanged(eventArgs:StateChangedEventArgs,_currentObject:any):void
  {
    if (eventArgs.OldValue != null && eventArgs.NewValue == null)
    {
      _currentObject.OnDependencyChanged(_currentObject);
    }
  }

  protected OnRunStarted():void{
    var selfDependencyFound=0;
    var Dependencies=this.GetDependencies();
    var OptionalDependencies=this.GetOptionalDependencies();
    for (var i = 0, len = Dependencies.length; i < len; i++) {
      if(Dependencies[i]==this._stateName)
      {
        selfDependencyFound=selfDependencyFound+1;
      }
      if(selfDependencyFound>0){ throw new Error('StateBuilder cannot have itself as a dependency.'); }
    }

    var selfOptionalDependencyFound=0;
    for (var j = 0, len = OptionalDependencies.length; j < len; j++) {
      if(OptionalDependencies[j]==this._stateName)
      {
        selfOptionalDependencyFound=selfOptionalDependencyFound+1;
      }
      if(selfOptionalDependencyFound>0){ throw new Error('StateBuilder cannot have itself as a dependency.'); }
    }

    this._workItem.SubscribeStateChangedEventWithArgs(this._stateName, this.OnStateChanged,this);//Need to check parameter here
    var anyDepedencyFound=0;
    for (var k = 0, len = Dependencies.length; k < len; k++) {
       this._workItem.SubscribeStateChangedEvent(Dependencies[k], this.OnDependencyChanged,this);
       anyDepedencyFound=anyDepedencyFound+1;
    }

    for (var l = 0, len = OptionalDependencies.length; l < len; l++) {
       this._workItem.SubscribeStateChangedEvent(OptionalDependencies[l], this.OnDependencyChanged,this);
       anyDepedencyFound=anyDepedencyFound+1;
    }
    if(anyDepedencyFound>0){
      this.OnDependencyChanged(this);
    }
  }

  public CreateStateBuilderContext(dependencies:Dictionary,optionalDependencies:Dictionary,_currentObject:any){
    return new StateBuilderCtx(_currentObject._workItem,_currentObject._stateName,_currentObject.GetDependencies(),_currentObject.GetOptionalDependencies(),null);
  }

  public OnDependencyChanged(_currentObject:any):void{
    var dependencies=_currentObject.GetDependencies();
    var optionalDependencies=_currentObject.GetOptionalDependencies();
    _currentObject._stateBuilderContext = _currentObject.CreateStateBuilderContext(dependencies, optionalDependencies,_currentObject);
    if(_currentObject._stateBuilderContext.Dependencies.length>0){
      _currentObject.OnDependenciesMet(_currentObject._stateBuilderContext,_currentObject);
    }
  }

  public OnDependenciesMet(stateBuilderContext:StateBuilderContext<any>,_currentObject:any):void{
    /**
      demo({ name: 'vijay' });
      function demo(pa: any): void {
          (async () => { await console.log('In async function'); console.log(pa); })();
      }
     */
    (async (_stateHandlerContext,_currentObj) => { 
        await _currentObj.DoWork(_stateHandlerContext,_currentObj);
      })(stateBuilderContext,_currentObject);
  }

  public async DoWork(stateBuilderContext:StateBuilderContext<any>,_currentObj:any){
    console.log(_currentObj);
    await _currentObj.GetStateObject(stateBuilderContext);
  }

  public abstract GetDependencies():Array<string>;
  public abstract GetOptionalDependencies():Array<string>;
  public abstract GetAdditionalData():Array<string>;

  public abstract GetStateObject(context:StateBuilderContext<any>):void;
}

class StateBuilderCtx extends StateBuilderContext<any>{
  public constructor(workItem:IWorkItem, stateName:string, dependencies:Dictionary, optionalDependencies:Dictionary, additionalData:Dictionary){
    super(workItem,stateName,dependencies,optionalDependencies,additionalData);
  }
}